### ISSUE TITLE
freeorion's compilation received out of memory

### ISSUE THREAD

**Issue by ghost** [freeorion/freeorion#3565]
Hi.
Trying to build freeorion with multiple processes (4, since my CPU is a simple dual-core multithreaded) with 15GiB of ram, I get processes killed because of lack of memory.
Freeorion was never light nor fast to build, but I think having compilation processes eating easily several GiB of VSZ or RSS is too much.
If I had to guess, I would suppose some header files include too much (boost libraries includes in headers?) and that this is the cause. There may even be some recursive includes or other things like that.

This is the 1st time I see a compilation processes eating that much memory.
For information:
`wc -c $( find \( -path '*build*' \) -prune -o -type f -a \( -name '*.h' -o -name '*.cpp' \)  )` gives ~14MiB. The same on other projects I compile routinely can go to more than 100MiB, and there's no RAM problem.

That's not blocking since I can still compile with a single process, but, I think this would be a good idea to try to improve the situation on this.

Additional informations:
Debian 11
ninja 1.10.1
clang++  11.0.1-2

**Comment by geoffthemedio** [comment]
It's probably not including too many headers. They aren't really that big, and they all have include guards against recursion. Rather, the issue is generally instantiating a lot of heavily nested template classes, particularly in the parser code. See this plot of my system's memory use when compiling just the parsers project (top) and just the common project (bottom) in MSVC2019:

![FO Building MSVC2017 Memory Usage Parsers vs Common](https://user-images.githubusercontent.com/11386908/135116908-de7532b6-1f86-45d1-a092-8d54e3980d64.png)

If you're running out of memory, see if you can avoid parallel builds of the parsers specifically, while keeping it for the rest of the code.

As for fixing it, it should be possible to (further) break up some of the parser code into smaller source files that each do less. Or, you could try to encourage someone (or help yourself) with migrating to the Python-parser based replacement that @o01eg was working on but which is currently seemingly stalled.

**Comment by ghost** [comment]
I'm just a user here. I am busy enough with other projects, sorry, all I can do is sending that bug reports (days only have a limited number of hours).

As for the build, using less processes of course worked, and I only had that problem because I recently moved the build folder so that my "contrib" folder is more organized and cmake does not like this (requires clean builds, since that tool stores full paths in the make/ninja files).

To me, this is not a critical problem nor even an important one, since it builds anyway. Slowly, but it does. Improving on it is only a matter of making development easier, it has no impact on usage.

**Comment by crabctrl** [comment]
Just wanted to mention, I have the same issue; all of the software I install is compiled from source, but FreeOrion is the *only* one I have to limit to one thread in order to avoid OOM, and even then I often need to enable swap... this memory usage is a bit excessive.

**Comment by Vezzra** [comment]
> this memory usage is a bit excessive

True. However, fixing this is going to be a long-term project, because the only way to fix it is replacing the boost parser library with something else. Which is already worked on (we intend to migrate the content scripts to Python), but it will take a lot of time and effort to finish this.

**Comment by daniel-santos** [comment]
@bmorel , @katp32 Welcome to [TMP](https://en.wikipedia.org/wiki/Template_metaprogramming)!  There's always a penalty somewhere when we use abstractions.  The awesome part about TMP is that we move the abstraction penalty to the compiler!  But the down side is that we move the abstraction penalty to the compiler.

**Comment by Vezzra** [comment]
> But the down side is that we move the abstraction penalty to the compiler.

In case of the boost parser libraries that penalty is unfortunately ridiculously excessive, in terms of resource usage during the build process (compile time and memory usage), bloated binaries, and in case of compile errors, completely unreadable and unusable error messages...

**Comment by daniel-santos** [comment]
> In case of the boost parser libraries that penalty is unfortunately ridiculously excessive, in terms of resource usage during the build process (compile time and memory usage), bloated binaries, and in case of compile errors, completely unreadable and unusable error messages...

Well, gcc has come a long way in making these diagnostics more readable and understandable, but ... uh yeah.  And I haven't even been doing much boost recently.  I'm not aware of binary bloat issues however.  I've been doing mostly C (kernel) and system libs the last few years though.
